Source/Dafny/Resolver.cs

```diff
@@ -3345,10 +3345,20 @@ public class ModuleBindings
               if (isString) return (string) e0 + (string) e1;
               break;
             case BinaryExpr.ResolvedOpcode.Sub:
-     //         if (isInt && (BigInteger) e0 == 0) return -(BigInteger) e1; // Allow this case for newtypes
               if (isInteger) return (BigInteger) e0 - (BigInteger) e1;
               if (isBV) return ((BigInteger) e0 - (BigInteger) e1) & MaxBV(bin.Type);
               if (isReal) return (Basetypes.BigDec) e0 - (Basetypes.BigDec) e1;
+              // Allow a special case: If the result type is a newtype that is integer-based (i.e., isInt && !isInteger)
+              // then we generally do not fold the operations, because we do not determine whether the
+              // result of the operation satsifies the new type constraint. However, on the occasion that
+              // a newtype aliases int witxhout a constraint, it occurs that a value of the newtype is initialized
+              // with a negative value, which is represented as "0 - N", that is, it comes to this case. It
+              // is a nuisance not to constant-fold the result, as not doing so can alter the determination
+              // of the representation type.
+              if (isInt && (BigInteger) e0 == 0 && bin.Type.AsNewtype != null &&
+                  bin.Type.AsNewtype.Constraint == null) {
+                return -(BigInteger) e1;
+              }
```

Comment:
Oh, I see. Yes, good one.

The base type of the newtype needs to be an unconstrained as well, to prevent something like this:

    newtype int8 = x | -128 <= x < 128
    newtype n8 = int8

where `bin.Type` on 3358 is `n8`.

It may be useful to have an `AsUnconstrainedNewtype` which is like `AsNewtype`, but it returns `null` if any ancestor base type has a constraint. That is, its base type either satisfies `.IsIntegerType` or satisfies `.AsUnconstrainedNewtype != null`.

Indeed, in this case, _any_ subtraction would be okay, not just when `(BigInteger)e0 == 0`. (Does this perhaps mean that the `isInteger` definition above could be set to `true` also for unconstrained newtypes? If so, then it would allow any arithmetic operations on those as well.

And the same idea could apply to reals as well

Btw, I like the contrast between `isReal` and `isAnyReal` better than the contrast between `isInteger` and `isInt`. Perhaps both pairs could be renamed to `isUnconstrainedInt` and `isConstrainedInt`, and analogously for real.

Reply:
Used isInt and isAnyInt. Generalized the check for constrained newtypes.